src/borg/item.pyx

```diff
@@ -528,35 +530,68 @@ class ItemDiff:
             return False
         return _chunk_content_equal(chunk_iterator1, chunk_iterator2)
 
+
+    @staticmethod
+    def _chunk_content_equal(chunks1, chunks2):
+        return  _chunk_content_equal(chunks1, chunks2)
     
+
+cdef inline size_t min_slice(size_t a, size_t b):
+    return a if a < b else b
+
+@cython.final
+cdef  class chunk_compare_state:
+    # TODO: remove the method table
+      cdef Py_ssize_t len 
+      cdef Py_ssize_t i 
+      cdef bytes buf  
+      cdef object chunks
+
+      def __cinit__(self, chunks):
+        self.len = 0
+        self.i = 0
+        self.chunks = chunks
+        self.advance(0)
+      
+      cdef  inline size_t slicelen(self):
+         return self.len - self.i
+
+      cdef  inline bint is_consumed(self):
+        return self.len == 0
+
+      cdef  inline void* current(self):
+        return (<char*> self.buf) + self.i
+
+      cdef  void advance(self, Py_ssize_t count):
+        self.i += count
+        assert self.i <= self.len
+        
+        if self.i == self.len:
+            self.buf = next(self.chunks, None)
+            if self.buf is None:
+                self.len=0
+            else:
+                self.len = len(self.buf)
+            self.i = 0
+        
+        
+
 cpdef _chunk_content_equal(chunks1, chunks2):
     """
     Compare chunk content and return True if they are identical.
 
     The chunks must be given as chunk iterators (like returned by :meth:`.DownloadPipeline.fetch_many`).
     """
 
-    end = object()
-    cdef size_t alen = 0, ai = 0
-    cdef size_t blen = 0, bi = 0
-
+    cdef chunk_compare_state s1=chunk_compare_state(chunks1), s2= chunk_compare_state(chunks2)
```

Comment:
?

Reply:
hose abstract the iteration state, that helps to make sense of the loop details